core/thread/inc/TPool.h

```diff
@@ -0,0 +1,204 @@
+#ifndef TPOOL_H_
+#define TPOOL_H_
+
+#include "TCollection.h"
+#include "TObjArray.h"
+#include "ROOT/TSeq.h"
+
+template<class subc>
+class TPool {
+public:
+   explicit TPool(){};
+   explicit TPool(size_t nThreads){};
+
+   // // Map
+   // //these late return types allow for a compile-time check of compatibility between function signatures and args,
+   // //and a compile-time check that the argument list implements a front() method (all STL sequence containers have it)
+   template<class F> auto Map(F func, unsigned nTimes) -> std::vector<decltype(func())>;
+   template<class F, class T> auto Map(F func, T &args) -> std::vector < decltype(++(args.begin()), args.end(), func(args.front())) >;
+   // /// \cond doxygen should ignore these methods
+   template<class F, class INTEGER> auto Map(F func, ROOT::TSeq<INTEGER> args) -> std::vector<decltype(func(args.front()))>;
+   template<class F> TObjArray Map(F func, TCollection &args);
+   template<class F, class T> auto Map(F func, std::initializer_list<T> args) -> std::vector<decltype(func(*args.begin()))>;
```

Comment:
My personal preference is to use `std::array` instead of `std::initializer_list`. The latter is okay for internals, but as a user interface, `std::array` is much much easier to "swallow".

I bet you're using `initializer_list` because template deduction doesn't work from initializer lists, right? :-) I think that suggests a design issue of this:

In principle, these `Map` functions take several `T` and yield a `U` for each `T. Maybe you could make that explicit through a `TMapper<T, U>`? That can then be constructed from a series of `T`s (or a ROOT collection that can then check the inheritance), a function taking `T` and returning `U` (use `std::function` for that) and giving a collection of `U`s (I'd just always return a `vector<U>` - much simpler.)

That also gets rid of the SFINAEs and the six overloads. What do you think?

I bet you can apply a similar pattern to MapReduce, too :-)